### DISCUSSION TITLE
sensitive vars in extensions

### DISCUSSION THREAD

**Discussion by malteludwig** [directus/directus#3759] [Archive]
I'm wondering what would be a good way to use sensitive information in an extension, like Api Keys etc?

When bundled in the extensions .js, they are exposed on the public login page.
I couldn't figure out if  there is a way to add vars to directus .env and access them from within an extension? 

Do you have any suggestions about this?

**Comment by rijkvanzanten** [comment]
You can add your own environment variables, and use them in serverside extensions like hooks/endpoints. I wouldn't recommend adding them into your client side extensions cause as you said they'd be publicly accessible. In case you need something private in an interface/module etc, i'd recommend adding both a custom endpoint _and_ the interface/module so you have a serverside point you can get the data from 👍🏻

**Reply by benhaynes** [reply]
This was added to the Docs.

**Reply by philleepflorence** [reply]
@benhaynes ... can you please reference the docs for this?

**Reply by benhaynes** [reply]
It's not up until the next release, but WILL be here: https://docs.directus.io/guides/extensions/accessing-data.html

**Reply by philleepflorence** [reply]
Much thanks!

**Reply by malteludwig** [reply]
@rijkvanzanten thanks! that's a good solution.

is it possible to limit permissions/access to custom endpoints?

**Reply by rijkvanzanten** [reply]
> is it possible to limit permissions/access to custom endpoints?

You can access the current user / role through `req.accountability`. At that point you can handle it however you want 🙂

**Reply by malteludwig** [reply]
sorry to bother, but how should a call to the custom endpoint from inside the extension look?

I can't figure out how to pass a token to the endpoint and so req.accountability.user / .role are null.
Is there a way to get the current users refresh token inside the extension and use something like 
`axios.get('/custom/custom-endpoint', {
                    headers: {
                        Authorization: `Bearer ${token}`,
                    },
                })`  
?

**Reply by philleepflorence** [reply]
@malteludwig - Are you following this and cannot get the information you seek: [Creating a Custom API Endpoint](https://docs.directus.io/guides/extensions/api-endpoints.html)?

Also, are you sending this request from a custom [module](https://docs.directus.io/guides/extensions/modules.html), [layout](https://docs.directus.io/guides/extensions/layouts.html), or [interface](https://docs.directus.io/guides/extensions/interfaces.html), while signed in to the Directus App?

**Reply by malteludwig** [reply]
@philleepflorence: I did follow this and searched the docs, but I'm stuck...

I'm sending the request from a custom module, logged in as admin.
the endpoint is working, just req.accountability is empty.

module.vue:
>const response = await axios.get('/custom/netlify-deploy');`

endpoint index.js:

> module.exports = function registerEndpoint(router, { env, services }) {
>      router.get('/', (req, res) => { 
>         console.log('req.accountability', req.accountability);
>       }); 
>};



output:
>req.accountability {
 >user: null,
 >role: null,
 >admin: false, 
> ...
> }

request: 
<img width="865" alt="Screenshot 2021-01-20 at 23 57 23" src="https://user-images.githubusercontent.com/40273858/105251252-6fd0ec00-5b7b-11eb-9974-a0d9139e2268.png">

**Reply by malteludwig** [reply]
I ended up parsing the cookie and requesting a refresh from AuthenticationService service,  then send the response if an access token is available and update the cookie with the new token.

Is that an acceptable way to do this?


module.exports = function registerEndpoint(router, { services, exceptions, env }) {
    const { AuthenticationService } = services;
    const { ServiceUnavailableException } = exceptions;

    router.get('/', async (req, res, next) => {
        var ms = require('ms');

        try {
            // check request cookie vor refresh token
            if (req.headers.cookie && req.headers.cookie.split('=')[0] === 'directus_refresh_token') {
                // request a refresh
                const authService = new AuthenticationService({
                    accountability: req.accountability,
                    schema: req.schema,
                });
                const auth = await authService.refresh(req.headers.cookie.split('=')[1]);

                if (auth.accessToken) {
                    // go on with the new token
                    const refreshToken = auth.refreshToken;

                    // store token in cookie
                    res.cookie('directus_refresh_token', refreshToken, {
                        httpOnly: true,
                        maxAge: ms(env.REFRESH_TOKEN_TTL),
                        secure: env.REFRESH_TOKEN_COOKIE_SECURE === 'true' ? true : false,
                        sameSite: env.REFRESH_TOKEN_COOKIE_SAME_SITE || 'strict',
                    });

                    // respond with vars
                    res.json({
                        netlifyClientID: env.NETLIFY_CLIENT_ID,
                        netlifySiteId: env.NETLIFY_SITE_ID,
                        netlifyTokenStorageId: env.NETLIFY_TOKEN_STORAGE_ID,
                    });
                }
                next();
            } else {
                // request has no cookie: request not authorized
                res.send('You must be authorized to do this.');
            }
        } catch (error) {
            // Passes errors into the error handler
            // res.send(e);
            return next(error);
        }
    });
};

**Reply by philleepflorence** [reply]
@rijkvanzanten - Please confirm. Also, let me know if the docs need updating to reflect this.

**Reply by malteludwig** [reply]
might be of help for others:

it turned out my error was to not inject `system` to module.vue

`export default {
    inject: ['system'],
...
}`

this allows to use  `system.api.get()`


then in endpoint.js `req.accountabilty` is filled.

**Reply by glowkeeper** [reply]
@malteludwig - I have a similar issue with accessing accountability data at a custom endpoint (see https://github.com/directus/directus/discussions/4551), but I don't quite understand your final answer. 

```
router.post(config.uRLs.resetPass.url, (req, res, next) => {

    const { error } = resetSchema.validate(req.body);
    if (error) return next(new InvalidPayloadException(error.message));

    console.log ("Accountability: ", req.accountability);
    if ( ( req.accountability?.admin !== true ) ||
         ( req.accountability?.user !== true ) ) {

        console.error(config.uRLs.resetPass.url + " unknown user");
        return next(new ServiceUnavailableException("Unavailable"));
    }

    const userid = req.body.userid;
    const newPassword = {
      password: req.body.newPass
    };

    const userService = new UsersService({ schema: req.schema });
    userService
      .update(newPassword, userid)
      .then( (results) => {

        return res.send(JSON.stringify(config.statusOkay));
      })
      .catch((error) => {

        console.error(config.uRLs.resetPass.url + " " + error.message)
        return next(new ServiceUnavailableException(error.message));
      });
  });
  
```

That all works beautifully _without_ the accountability guard, but with it, _req.accountability_ is empty, even with a valid token passed in the header. How did you fix your similar issue?

**Comment by malteludwig** [comment]
Great, thanks for your help!
